### ISSUE TITLE
Support for GPFlow as a backend

### ISSUE THREAD

**Issue by kiudee** [kiudee/bayes-skopt#104]
Currently, the library uses scikit-optimize/sklearn as a backend to do Gaussian process computations. These implementations are easy to use and served the library well so far.
A big problem is that the library is quite limited in functionality. One big use case of bayes-skopt is to handle very noisy target functions. In that regard it would be useful to be able to model heteroscedastic noise, which the Gaussian process in sklearn does not really support. It is possible to set the `alpha` parameter to a vector, which will incorporate the noise of the training data, but during prediction it is still noiseless.
This could be useful for acquisition functions which properly handle the observation noise, like noisy EI and knowledge gradient.

GPFlow offers a lot of the needed functionality out of the box. It is straightforward to construct heteroscedastic likelihoods. It also supports stochastic variational Gaussian processes, which allow GPs to scale to more than 10k observations.
Therefore, migrating to GPflow as a backend would be a good long-term goal. Sadly, it will require a major rewrite of the library, since many classes (Optimizer, acquisition functions etc.) are tightly coupled to the current GP implementation.

Roadmap
----------

- [ ] Get familiar with all relevant aspects of GPFlow.
- [ ] Identify an interface for the GP component which would allow plugging in different backends.
- [ ] Migrate the existing code to obey the new interface specification, reducing the coupling between the components.
- [ ] Add the GPFlow GP implementation.
- [ ] Add GPFlow specific features.
